### DISCUSSION TITLE
PWA (progressive web apps) support

### DISCUSSION THREAD

**Discussion by ghost** [Floorp-Projects/Floorp#411] [Feature Request]
### Is your feature request related to a problem?

Yes, there is no PWA support for Firefox by default. There are many sites I install as PWA, and that's why I go with a Chromium browser. There is an extension for pwa support, but it requires you to install some other things as well. I'd rather have it built into the browser. I think native PWA support will put Floorp ahead of firefox and other forks.

### Describe the solution you'd like

Native PWA support for Floorp

**Comment by Ryosuke-Asano** [comment]
I don't have time to work on it, so I can't guarantee it, but it is among the features we want to add it.

**Comment by dominichayesferen** [comment]
There is an upstream effort in the form of the Firefox PWAs Project, so their work may be a good reference point for starting implementation.

**Comment by Ryosuke-Asano** [comment]
> There is an upstream effort in the form of the Firefox PWAs Project, so their work may be a good reference point for starting implementation.

I have an idea. Use another approach is good for native implement.

**Comment by gunir** [comment]
Would be wonderful to be able to bring back PWA, it's very useful in my opinion:
- Reduce burden for main browser
- Improve privacy because PWA application is isolated from main profile, they can't get a lot of your 3rd party cookies
- Easy management

This project has a lot of interesting features: https://github.com/filips123/PWAsForFirefox
- Profile template allowing user to create new PWA with preconfigured `about:config`'s settings
- Auto open in PWA, for example you created a Netflix PWA, then if you open Netflix in your Floorp it automatically open the Netflix PWA

**Comment by black7375** [comment]
Another PWA implementation that I've had a good experience with is [Web catalog](https://webcatalog.io/).
It's [moved to closed source now](https://github.com/webcatalog/webcatalog-legacy), but it's license-compatible with MPL 2.0, so it's worth a look on the frontend.

**Comment by Ryosuke-Asano** [comment]
Hey. Big news.

Currently, I'm implementing PWA to Floorp. On Windows, PWA will be available on 11.7.0 or later.

https://github.com/Floorp-Projects/Floorp/assets/73892113/fc39a70a-5e76-48e4-850d-3d70ec5a514b

**Reply by gunir** [reply]
Beauty!

**Reply by black7375** [reply]
Looks good!!

**Reply by Ryosuke-Asano** [reply]
New interface of PWA (Lepton)
![image](https://github.com/Floorp-Projects/Floorp/assets/73892113/906d0e6b-87d2-4f40-abae-d707d6c7a17f)
![image](https://github.com/Floorp-Projects/Floorp/assets/73892113/f700c4e4-cdcc-4427-b138-0811351d2ed9)

**Reply by tigexplorer** [reply]
Thanx for your work!
Is it hard to implement it for Linux too?

**Reply by Ryosuke-Asano** [reply]
On Linux, API is missing...

**Reply by gunir** [reply]
We might need to ask public for help at this point, we have a lot of Mac users, which is very surprising it would be great to be able to implement very basic PWA without shortcut/pinning instead.

**Reply by p134c0d3** [reply]
@surapunoyousei as someone who *almost* exclusively uses Linux, this would be amazing. What is the project name of the API? I can go make a request there. 🙂 PWA support was the only thing keeping me on Brave tbh. Love the Firefox base. If they can get an API for *nix I'd switch everything over from Brave to Floorp. Love the browser and customizations that come with it!

**Reply by Ryosuke-Asano** [reply]
So.. API means Firefox's inside API. I create feature with JavaScript. Mozilla does not provide it.

**Reply by p134c0d3** [reply]
Thanks for the clarification. 🙂

**Comment by Ryosuke-Asano** [comment]
We finished Windows PWA implementation. This will be available at Floorp 11.7.0. I'm so exciting to try this feature for users! 

Floorp has THE BIG BANG now.

Thanks @black7375 for Lepton's icon PWA implements!

![image](https://github.com/Floorp-Projects/Floorp/assets/73892113/471bf1fb-1cb7-405b-899b-3d649db7e80b)
![image](https://github.com/Floorp-Projects/Floorp/assets/73892113/caa7aeb6-cccf-4966-a2ac-6d7409679dc2)
![image](https://github.com/Floorp-Projects/Floorp/assets/73892113/6164c3b9-1b0c-4df0-8a20-1824b841d944)


See you again at Floorp 11.7.0

**Reply by gunir** [reply]
Big feature, congrats this is amazing! 

Thanks for your hardwork!

**Reply by black7375** [reply]
Nice feature!!
It's a prime example of the benefits of a forked browser.

**Comment by ghost** [comment]
That's great, good job :)

**Comment by ghost** [comment]
@surapunoyousei Hi, first of all thanks for adding the PWA feature. It is great to have PWA as native. Something to note though, can you configure the PWA so that only the top bar is visible with no url bar (like on chromium PWAs, you can still access extensions and others with 3 dots). So these PWAs would look cleaner and more like native apps than a browser window. I added a picture for comparison.

![image](https://github.com/Floorp-Projects/Floorp/assets/140727743/22c3f583-715c-43a4-89f1-344095b38552)

**Reply by Ryosuke-Asano** [reply]
Difficulrt....

**Reply by ghost** [reply]
I see, maybe in the future then :)

**Reply by Void48** [reply]
I would also appreciate this a lot. Hope you can give it a try sometimes!

**Comment by gunir** [comment]
I wonder if it's possible to open PWA in Container, that allows us to use multiple accounts in PWA instead of using account from Default Container.

**Reply by Sohail-XD** [reply]
This is the whole point of a web app, the current implementation is simply a website shortcut added to os applauncher. 

Hope there would be container like system where one could change the firefox account/extenstions irrespective of the floorp browser.

**Comment by Firehawke** [comment]
How do you generate a start menu link for an installed PWA? I can see how to actually add the PWA, but it's not automatically generating a start menu link like Chromium would.

**Reply by H4M5TER** [reply]
In my case it does generate start menu entry for me.
There was a pop-up from windows, and after clicking yes there will be both link on start menu and taskbar.
`Windows 11 23H2 10.0.22631.6199`

**Reply by Firehawke** [reply]
Hm. Weird. Not doing it here despite the popup. It shows up in the PWA list in Floorp, but no start menu icon is created.

**Comment by H4M5TER** [comment]
Is there a way to edit the url of web app after creation?
Neither `Menu - Web App` nor `about:hub#/features/webapps` can do this for me.
`Floorp Version 12.15.3@152.0.5`
